torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -328,117 +295,102 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
-            state_dict = self.local_state_dict()
-            state_dict["local_state_dict"] = True
-            return state_dict
-
-        # Flatten the param_groups, save the partition which logs the rank <> shard correspondence
-        partition: List[Tuple[int, int]] = []
-        param_groups: List[Dict[Any, Any]] = []
-
-        start = 0
-        for i, s in enumerate(self._all_states):
-            param_groups.extend(s["param_groups"])
-            end = start + len(s["param_groups"])
-            partition.append((start, end))
-            start = end
-
-        return {
-            "state": [s["state"] for s in self._all_states],
-            "param_groups": param_groups,
-            "partition": partition,
-            "local_state_dict": False,
-        }
-
-    @staticmethod
-    def rank_local_state_dict(rank: int, state_dict: dict) -> dict:
-        """Returns the local_state_dict for a given rank.
+            raise RuntimeError(
+                "Optimizer state has not been consolidated on this rank. \
+                Please call `consolidate_state_dict()` on all ranks beforehand if you meant to save the global state"
+            )
+
+        # Unify the shard states and the state that pytorch would expect, given the model.
+        # Indexation needs several redirections, since each shard only knows a limited scope of the model
+        # - get the pytorch compliant parameter indexing
+        state_dict = super().state_dict()
+
+        # - get an id map which links the parameter id to the index in the reference state
+        global_id_map = {id(p): i for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        # - go through the per-shard states, which are all indexed locally
+        unordered_state = {}
+        for rank, s in enumerate(self._all_states):
+            # -- match the local indexing and the global partition, update the corresponding saved state globally
+            for local_pg, global_pg in zip(s["param_groups"], self.partition_parameters()[rank]):
+                # Go through the parameters indexed locally, pick up the global corresponding param
+                # NOTE: Contents of the state_dict changes in between torch1.5 and torch1.6+
+                local_index_to_param_id = {
+                    i_param: id(global_pg["params"][i]) for i, i_param in enumerate(local_pg["params"])
+                }
+
+                for local_param_index in local_pg["params"]:
+                    # Update the state, if any
+                    if local_param_index in s["state"].keys():
+                        unordered_state[local_index_to_param_id[local_param_index]] = s["state"][local_param_index]
+
+        # - save the states in the expected order, meaning the one from the param_groups
+        for i, param in enumerate(chain(*(g["params"] for g in self.param_groups))):
+            if id(param) in unordered_state.keys():
```

Comment:
when will a parameter not included in `unordered_state`?

Reply:
sgd without momentum for instance, in that case there's just no state. I'm not 100% sure, but also if a param is not trainable, my guess is that it would show up in params group but not in the optimizer state ?